File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/homes/HomeTaskRecurrenceDtos.kt
package app.pantopus.android.data.api.models.homes

import app.pantopus.android.data.homes.homeTaskUUID
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.time.Instant
import java.time.ZoneId

internal const val MAX_RECURRENCE_REVISION = 999_999_999_999_999L
internal val RECURRENCE_FREQUENCIES = setOf("DAILY", "WEEKLY", "MONTHLY")

internal fun recurrenceDate(value: String?): Instant? = value?.let { runCatching { Instant.parse(it) }.getOrNull() }

internal fun recurrenceZone(value: String?): Boolean = value != null && (value == "UTC" || value in ZoneId.getAvailableZoneIds())

@JsonClass(generateAdapter = true)
data class HomeTaskAutomaticRecurrence(
    val state: String,
    val frequency: String,
    val interval: Int,
    val timezone: String,
    @Json(name = "next_due_at") val nextDueAt: String? = null,
) {
    fun valid(): Boolean =
        state in setOf("active", "paused", "needs_review") && frequency in RECURRENCE_FREQUENCIES &&
            interval in 1..365 && recurrenceZone(timezone) &&
            (if (state == "active") recurrenceDate(nextDueAt) != null else nextDueAt == null)

    fun label(): String =
        when (state) {
            "active" -> "Repeats every $interval ${recurrencePeriod(frequency)}${if (interval == 1) "" else "s"}"
            "paused" -> "Repeats paused"
            else -> "Repeats need review"
        }
}

internal fun recurrencePeriod(frequency: String): String =
    when (frequency) {
        "DAILY" -> "day"
        "WEEKLY" -> "week"
        else -> "month"
    }

@JsonClass(generateAdapter = true)
data class HomeTaskRecurrenceConfiguration(
    val id: String,
    val revision: Long,
    val state: String,
    val frequency: String,
    val interval: Int,
    val timezone: String,
    @Json(name = "anchor_at") val anchorAt: String,
    @Json(name = "generated_count") val generatedCount: Long,
    val reason: String? = null,
    @Json(name = "next_due_at") val nextDueAt: String? = null,
    @Json(name = "last_due_at") val lastDueAt: String? = null,
    @Json(name = "last_task_id") val lastTaskId: String? = null,
) {
    fun valid(): Boolean =
        homeTaskUUID(id) && revision in 1..MAX_RECURRENCE_REVISION && generatedCount >= 0 &&
            recurrenceDate(anchorAt) != null && (lastDueAt == null || recurrenceDate(lastDueAt) != null) &&
            (lastTaskId == null || homeTaskUUID(lastTaskId)) &&
            HomeTaskAutomaticRecurrence(state, frequency, interval, timezone, nextDueAt).valid()
}

/** GET and POST share the current projection; only POST adds its original receipt. */
@JsonClass(generateAdapter = true)
data class HomeTaskRecurrenceState(
    val ok: Boolean,
    @Json(name = "home_id") val homeId: String,
    @Json(name = "task_id") val taskId: String,
    @Json(name = "can_manage") val canManage: Boolean,
    @Json(name = "task_updated_at") val taskUpdatedAt: String,
    val revision: Long,
    @Json(name = "task_session") val taskSession: HomeTaskSessionDto,
    val configuration: HomeTaskRecurrenceConfiguration? = null,
    val receipt: HomeTaskRecurrenceReceipt? = null,
    val replayed: Boolean? = null,
) {
    fun matches(
        home: String,
        task: String,
    ): Boolean =
        ok && homeId == home && taskId == task && recurrenceDate(taskUpdatedAt) != null &&
            revision in 0..MAX_RECURRENCE_REVISION &&
            (configuration?.let { it.valid() && it.revision == revision } ?: (revision == 0L))
}

/** This is the exact immutable wire command. Null start-only fields are omitted. */
@JsonClass(generateAdapter = true)
data class HomeTaskRecurrenceRequest(
    @Json(name = "request_id") val requestId: String,
    val action: String,
    @Json(name = "expected_revision") val expectedRevision: Long,
    @Json(name = "expected_task_updated_at") val expectedTaskUpdatedAt: String? = null,
    val frequency: String? = null,
    val interval: Int? = null,
    val timezone: String? = null,
) {
    fun valid(): Boolean {
        if (!homeTaskUUID(requestId) || expectedRevision !in 0..MAX_RECURRENCE_REVISION) return false
        if (action == "pause") return expectedTaskUpdatedAt == null && frequency == null && interval == null && timezone == null
        return action == "start" && recurrenceDate(expectedTaskUpdatedAt) != null && frequency in RECURRENCE_FREQUENCIES &&
            interval != null && interval in 1..365 && recurrenceZone(timezone)
    }
}

@JsonClass(generateAdapter = true)
data class HomeTaskRecurrenceReceipt(
    @Json(name = "request_id") val requestId: String,
    @Json(name = "actor_id") val actorId: String,
    @Json(name = "home_id") val homeId: String,
    @Json(name = "task_id") val taskId: String,
    val action: String,
    val revision: Long,
    @Json(name = "request_hash") val requestHash: String,
    @Json(name = "created_at") val createdAt: String,
) {
    fun matches(request: HomeTaskRecurrenceRequest): Boolean {
        val expected = request.expectedRevision + if (request.action == "start" || request.expectedRevision > 0) 1 else 0
        return requestId == request.requestId && action == request.action && revision == expected &&
            requestHash.matches(Regex("^[a-f0-9]{64}$")) && recurrenceDate(createdAt) != null
    }
}

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Homes/HomeTaskRecurrenceDTOs.swift
import Foundation

public enum HomeTaskRecurrenceFrequency: String, Codable, Sendable, CaseIterable {
    case daily = "DAILY"
    case weekly = "WEEKLY"
    case monthly = "MONTHLY"

    var period: String {
        switch self {
        case .daily: "day"
        case .weekly: "week"
        case .monthly: "month"
        }
    }
}

enum HomeTaskRecurrenceDate {
    static func parse(_ value: String?) -> Date? {
        guard let value else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: value) { return date }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: value)
    }

    static func label(_ value: String, timezone: String) -> String {
        guard let date = parse(value), let zone = TimeZone(identifier: timezone) else { return "Date unavailable" }
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        formatter.timeZone = zone
        return formatter.string(from: date)
    }
}

public struct HomeTaskAutomaticRecurrence: Decodable, Sendable, Hashable {
    public let state: String
    public let frequency: HomeTaskRecurrenceFrequency
    public let interval: Int
    public let timezone: String
    public let nextDueAt: String?

    private enum CodingKeys: String, CodingKey {
        case state, frequency, interval, timezone
        case nextDueAt = "next_due_at"
    }

    var valid: Bool {
        ["active", "paused", "needs_review"].contains(state) && (1...365).contains(interval)
            && TimeZone(identifier: timezone) != nil
            && (state == "active" ? HomeTaskRecurrenceDate.parse(nextDueAt) != nil : nextDueAt == nil)
    }

    var label: String {
        switch state {
        case "active": "Repeats every \(interval) \(frequency.period)\(interval == 1 ? "" : "s")"
        case "paused": "Repeats paused"
        default: "Repeats need review"
        }
    }
}

struct HomeTaskRecurrenceConfiguration: Decodable, Equatable {
    let id: String
    let revision: Int
    let state: String
    let reason: String?
    let frequency: HomeTaskRecurrenceFrequency
    let interval: Int
    let timezone: String
    let anchorAt: String
    let nextDueAt: String?
    let lastDueAt: String?
    let lastTaskId: String?
    let generatedCount: Int

    private enum CodingKeys: String, CodingKey {
        case id, revision, state, reason, frequency, interval, timezone
        case anchorAt = "anchor_at", nextDueAt = "next_due_at", lastDueAt = "last_due_at"
        case lastTaskId = "last_task_id", generatedCount = "generated_count"
    }

    var valid: Bool {
        UUID(uuidString: id) != nil && revision > 0 && revision <= 999_999_999_999_999
            && ["active", "paused", "needs_review"].contains(state)
            && (1...365).contains(interval) && TimeZone(identifier: timezone) != nil
            && HomeTaskRecurrenceDate.parse(anchorAt) != nil && generatedCount >= 0
            && (state == "active" ? HomeTaskRecurrenceDate.parse(nextDueAt) != nil : nextDueAt == nil)
            && (lastDueAt == nil || HomeTaskRecurrenceDate.parse(lastDueAt) != nil)
            && (lastTaskId == nil || lastTaskId.flatMap(UUID.init(uuidString:)) != nil)
    }
}

struct HomeTaskRecurrenceState: Decodable {
    let ok: Bool
    let homeId: String
    let taskId: String
    let canManage: Bool
    let taskUpdatedAt: String
    let revision: Int
    let configuration: HomeTaskRecurrenceConfiguration?
    let taskSession: HomeTaskSession

    private enum CodingKeys: String, CodingKey {
        case ok, revision, configuration
        case homeId = "home_id", taskId = "task_id", canManage = "can_manage"
        case taskUpdatedAt = "task_updated_at", taskSession = "task_session"
    }

    func matches(home: String, task: String) -> Bool {
        ok && homeId == home && taskId == task && HomeTaskRecurrenceDate.parse(taskUpdatedAt) != nil
            && revision >= 0 && revision <= 999_999_999_999_999
            && (configuration.map { $0.valid && $0.revision == revision } ?? (revision == 0))
    }
}

struct HomeTaskRecurrenceResponse: Decodable {
    let state: HomeTaskRecurrenceState
    let receipt: HomeTaskRecurrenceReceipt
    let replayed: Bool

    private enum CodingKeys: CodingKey { case receipt, replayed }

    init(from decoder: any Decoder) throws {
        state = try HomeTaskRecurrenceState(from: decoder)
        let fields = try decoder.container(keyedBy: CodingKeys.self)
        receipt = try fields.decode(HomeTaskRecurrenceReceipt.self, forKey: .receipt)
        replayed = try fields.decode(Bool.self, forKey: .replayed)
    }
}
